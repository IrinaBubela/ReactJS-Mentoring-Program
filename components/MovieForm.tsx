import React from 'react';
import { Form, Field, ErrorMessage, Formik, FormikHelpers, FormikErrors } from 'formik';
import * as Yup from 'yup';
import { Movie, emptyMovieForPost } from './MovieListPage';
interface MovieFormProps {
  onSubmit: (formData: Movie) => void;
  initialInfo?: Movie;
}

const validationSchema = Yup.object().shape({
  title: Yup.string().required('Title is required'),
  tagline: Yup.string(),
  vote_average: Yup.number().min(0, 'Vote average must be a positive number'),
  vote_count: Yup.number().min(0, 'Vote count must be a positive number'),
  release_date: Yup.string().required('Release Date is required'),
  poster_path: Yup.string().url().required('Invalid URL format'),
  overview: Yup.string().required('Overview is required'),
  budget: Yup.number().min(0, 'Budget must be a positive number'),
  revenue: Yup.number().min(0, 'Revenue must be a positive number'),
  runtime: Yup.number().required('Runtime is required').positive('Runtime must be a positive number'),
  genres: Yup.array().of(Yup.string()).min(1, 'At least one genre is required'),
});

const MovieForm: React.FC<MovieFormProps> = ({ onSubmit, initialInfo }) => {
  const handleGenresChange =
    (event: Event,
      setFieldValue: (field: string, value: any, shouldValidate?: boolean | undefined) => Promise<void | FormikErrors<Movie>>
    ) => {
      const inputValue = event.target as HTMLInputElement;
      const genresArray = inputValue.value.split(',').map(genre => genre.trim());
      setFieldValue('genres', genresArray);
    }

  const handleSubmit = (values: Movie, { setSubmitting }: FormikHelpers<any>) => {
    setTimeout(() => {
      alert(`New Film ${values.title} was added`);
      setSubmitting(false);
    }, 400);
    onSubmit(values);
  };

  return (
    <Formik
      initialValues={initialInfo || emptyMovieForPost}
      validationSchema={validationSchema}
      onSubmit={handleSubmit}
    >
      {({ isSubmitting, setFieldValue }) => (
        <Form>
        <div className="row">
          <div className="col-8">
            <label htmlFor="title" className="form-label">Title</label>
            <Field type="text" className="form-control" id="title" name="title" />
            <ErrorMessage name="title" component="div" className="text-danger" />
          </div>
          <div className="col-4">
            <label htmlFor="release_date" className="form-label">Release Year</label>
            <Field type="text" className="form-control" id="release_date" name="release_date" />
            <ErrorMessage name="release_date" component="div" className="text-danger" />
          </div>
      
          <div className="col-8">
            <label htmlFor="poster_path" className="form-label">Movie URL</label>
            <Field type="text" className="form-control" id="poster_path" name="poster_path" />
            <ErrorMessage name="poster_path" component="div" className="text-danger" />
          </div>
      
          <div className="col-8">
            <label htmlFor="genres">Genres (comma-separated)</label>
            <Field type="text" className="form-control" id="genres" name="genres" onChange={(event: Event) => handleGenresChange(event, setFieldValue)} />
            <ErrorMessage name="genres" component="div" className="text-danger" />
          </div>
          <div className="col-4">
            <label htmlFor="runtime" className="form-label">Runtime</label>
            <Field type="text" className="form-control" id="runtime" name="runtime" />
            <ErrorMessage name="runtime" component="div" className="text-danger" />
          </div>
      
          <div className="col-12">
            <label htmlFor="overview" className="form-label">Overview</label>
            <Field as="textarea" className="form-control" id="overview" name="overview" />
            <ErrorMessage name="overview" component="div" className="text-danger" />
          </div>
      
          {/* New Fields */}
          <div className="col-8">
            <label htmlFor="tagline" className="form-label">Tagline</label>
            <Field type="text" className="form-control" id="tagline" name="tagline" />
            <ErrorMessage name="tagline" component="div" className="text-danger" />
          </div>
          <div className="col-4">
            <label htmlFor="vote_average" className="form-label">Vote Average</label>
            <Field type="text" className="form-control" id="vote_average" name="vote_average" />
            <ErrorMessage name="vote_average" component="div" className="text-danger" />
          </div>
      
          <div className="col-8">
            <label htmlFor="budget" className="form-label">Budget</label>
            <Field type="text" className="form-control" id="budget" name="budget" />
            <ErrorMessage name="budget" component="div" className="text-danger" />
          </div>
          <div className="col-4">
            <label htmlFor="revenue" className="form-label">Revenue</label>
            <Field type="text" className="form-control" id="revenue" name="revenue" />
            <ErrorMessage name="revenue" component="div" className="text-danger" />
          </div>
      
          <div className="col-12">
            <label htmlFor="vote_count" className="form-label">Vote Count</label>
            <Field type="text" className="form-control" id="vote_count" name="vote_count" />
            <ErrorMessage name="vote_count" component="div" className="text-danger" />
          </div>
        </div>
        <div className="form-actions">
          <div className="form-actions-btns">
            <div className="text-end mt-3">
              <button type="submit" className="btn btn-primary" disabled={isSubmitting}>Submit</button>
            </div>
          </div>
        </div>
      </Form>
      
      )}
    </Formik >
  );
};

export default MovieForm;
