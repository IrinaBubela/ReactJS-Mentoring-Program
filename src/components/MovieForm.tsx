import React, { useState } from 'react';
import { initialMovieInfoObject } from '../App';

interface MovieInfo {
  title: string;
  releaseYear: string;
  imageUrl: string;
  rating: string;
  genres: string[];
  duration: string;
  overview: string;
}

interface MovieFormProps {
  initialMovieInfo?: MovieInfo;
  onSubmit: (formData: MovieInfo) => void;
}

const MovieForm: React.FC<MovieFormProps> = ({ initialMovieInfo, onSubmit }) => {
  const [movieInfo, setMovieInfo] = useState<MovieInfo>(initialMovieInfo || initialMovieInfoObject);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const { name, value } = e.target;
    setMovieInfo({ ...movieInfo, [name]: value });
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    onSubmit(movieInfo);
  };

  return (
    <form className='container' onSubmit={handleSubmit}>
      <div className="row">
        <div className="col-8">
          <label htmlFor="title" className="form-label">Title</label>
          <input type="text" className="form-control" id="title" name="title" value={movieInfo.title} onChange={handleChange} />
        </div>
        <div className="col-4">
          <label htmlFor="releaseYear" className="form-label">Release Year</label>
          <input type="text" className="form-control" id="releaseYear" name="releaseYear" value={movieInfo.releaseYear} onChange={handleChange} />
        </div>

        <div className="col-8">
          <label htmlFor="imageUrl" className="form-label">Movie URL</label>
          <input type="text" className="form-control" id="imageUrl" name="imageUrl" value={movieInfo.imageUrl} onChange={handleChange} />
        </div>
        <div className="col-4">
          <label htmlFor="rating" className="form-label">Rating</label>
          <input type="text" className="form-control" id="rating" name="rating" value={movieInfo.rating} onChange={handleChange} />
        </div>

        <div className="col-8">
          <label htmlFor="genres" className="form-label">Genre</label>
          <input type="text" className="form-control" id="genres" name="genres" value={movieInfo.genres} onChange={handleChange} />
        </div>
        <div className="col-4">
          <label htmlFor="duration" className="form-label">Runtime</label>
          <input type="text" className="form-control" id="duration" name="duration" value={movieInfo.duration} onChange={handleChange} />
        </div>

        <div className="col-12">
          <label htmlFor="overview" className="form-label">Overview</label>
          <textarea className="form-control" id="overview" name="overview" value={movieInfo.overview} onChange={handleChange}></textarea>
        </div>
      </div>
      <div className="modal-footer">
        <div className="text-end mt-3">
          <button type="button" className="btn btn-primary">Reset</button>
        </div>
        <div className="text-end mt-3">
          <button type="submit" className="btn btn-primary">Submit</button>
        </div>
      </div>
    </form>
  );
};

export default MovieForm;
