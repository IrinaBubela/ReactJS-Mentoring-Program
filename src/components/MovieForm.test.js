import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import MovieForm from './MovieForm';

describe('MovieForm Component', () => {
  const initialMovieInfo = {
    title: 'Test Movie',
    releaseYear: '2022',
    imageUrl: 'https://example.com/movie.jpg',
    rating: 'PG-13',
    genres: 'Action, Adventure',
    duration: '120',
    overview: 'This is a test movie.',
  };

  it('renders movie form with initial values', () => {
    const onSubmit = jest.fn();
    const { asFragment, getByLabelText, getByText } = render(
      <MovieForm initialMovieInfo={initialMovieInfo} onSubmit={onSubmit} />
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(getByText('Submit'));
    expect(onSubmit).toHaveBeenCalledWith(initialMovieInfo);
  });
});
