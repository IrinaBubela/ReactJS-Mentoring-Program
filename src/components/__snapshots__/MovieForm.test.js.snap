// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm Component renders movie form with initial values 1`] = `
<DocumentFragment>
  <form
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <label
          class="form-label"
          for="title"
        >
          Title
        </label>
        <input
          class="form-control"
          id="title"
          name="title"
          type="text"
          value="Test Movie"
        />
      </div>
      <div
        class="col-4"
      >
        <label
          class="form-label"
          for="releaseYear"
        >
          Release Year
        </label>
        <input
          class="form-control"
          id="releaseYear"
          name="releaseYear"
          type="text"
          value="2022"
        />
      </div>
      <div
        class="col-8"
      >
        <label
          class="form-label"
          for="imageUrl"
        >
          Movie URL
        </label>
        <input
          class="form-control"
          id="imageUrl"
          name="imageUrl"
          type="text"
          value="https://example.com/movie.jpg"
        />
      </div>
      <div
        class="col-4"
      >
        <label
          class="form-label"
          for="rating"
        >
          Rating
        </label>
        <input
          class="form-control"
          id="rating"
          name="rating"
          type="text"
          value="PG-13"
        />
      </div>
      <div
        class="col-8"
      >
        <label
          class="form-label"
          for="genres"
        >
          Genre
        </label>
        <input
          class="form-control"
          id="genres"
          name="genres"
          type="text"
          value="Action, Adventure"
        />
      </div>
      <div
        class="col-4"
      >
        <label
          class="form-label"
          for="duration"
        >
          Runtime
        </label>
        <input
          class="form-control"
          id="duration"
          name="duration"
          type="text"
          value="120"
        />
      </div>
      <div
        class="col-12"
      >
        <label
          class="form-label"
          for="overview"
        >
          Overview
        </label>
        <textarea
          class="form-control"
          id="overview"
          name="overview"
        >
          This is a test movie.
        </textarea>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <div
        class="text-end mt-3"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Reset
        </button>
      </div>
      <div
        class="text-end mt-3"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
